### PR TITLE
Allow integration of libOQS

### PR DESCRIPTION
**Pull request by brett-warren-arm**
This PR adds the configuration flag `MBEDTLS_LIBOQS_ENABLE` which controls the integration of the libOQS post-quantum cryptography library into Mbed TLS.

### CODE REVIEW

**Comment by hanno-becker** [comment]
Hi @brett-warren-arm, thanks! Can you add a brief description of what's going on the in the PR?

**Comment by brett-warren-arm** [comment]
**As described in the commit message:**

SSUE: liboqs currently has to be bulit first manually:

	ninja liboqs
	(then whatever commands you desired after)

This is because liboqs.a has to exist, otherwise cmake will
complain about it not existing.

**To elaborate:**

This is where the library is linked,

```
    if(TARGET liboqs)
        add_dependencies(${mbedcrypto_target} PUBLIC liboqs)
	target_link_libraries(${mbedcrypto_target} PUBLIC ${CMAKE_BINARY_DIR}/lib/liboqs.a)
    endif()
```

cmake complains about liboqs.a not existing, before the liboqs project can be built. I don't know
if there's another way of expressing this so that liboqs.a's existence is checked later, or perhaps
some other way of making this behave correctly.

**Comment by hannestschofenig** [comment]
Adding the liboqs.a library alone will not be enough to add support for PQC algorithms to the TLS 1.3 implementation. Do you agree?

**Comment by brett-warren-arm** [comment]
Yes, a new module will need to be added to mbedtls to integrate the PQC algorithms into the key schedule.

**Comment by hanno-becker** [comment]
@hannestschofenig We're meeting once a week to discuss the steps ahead - do you want to join?

File: 3rdparty/liboqs/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+include(ExternalProject)
+ExternalProject_Add(liboqs
+	GIT_REPOSITORY             https://github.com/open-quantum-safe/liboqs
+	GIT_TAG                    main
+	CMAKE_CACHE_ARGS           -DOQS_USE_OPENSSL:BOOL=OFF -DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_BINARY_DIR}
```
Comment (hanno-becker) [inline 3rdparty/liboqs/CMakeLists.txt]:
```suggestion
	CMAKE_CACHE_ARGS           -DOQS_ENABLE_KEM_CLASSIC_MCELIECE:BOOL=OFF -DOQS_USE_OPENSSL:BOOL=OFF -DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_BINARY_DIR}
```
Adding that to resolve some compilation issues, which at least I have trying to build classic McEliece for AVX2.

File: programs/CMakeLists.txt [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
-endif()
-add_subdirectory(hash)
-add_subdirectory(pkey)
-add_subdirectory(psa)
-add_subdirectory(random)
-add_subdirectory(ssl)
-add_subdirectory(test)
-add_subdirectory(util)
-add_subdirectory(x509)
+#add_subdirectory(aes)
```
Comment (hanno-becker) [inline programs/CMakeLists.txt]: Why were those things commented out?

File: programs/ssl/CMakeLists.txt [hunk 3]
```diff
@@ -30,11 +28,6 @@ foreach(exe IN LISTS executables)
     target_include_directories(${exe} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/../../tests/include)
 endforeach()
 
-set_property(TARGET ssl_client2 APPEND PROPERTY SOURCES
```
Comment (hanno-becker) [inline programs/ssl/CMakeLists.txt]: Why?

**Review by hanno-becker — CHANGES_REQUESTED** [review]
@brett-warren-arm The build is failing when `MBEDTLS_LIBOQS_ENABLED` is unset. It looks like the addition of liboqs as a library dependency should be conditional on that option being set. Can you fix that.

Also, something seems wrong with checking whether a re-build of liboqs is necessary -- on my machine, every `make` triggers an entire rebuild. Can you also look into that?

Otherwise, the PR is good to go.

**Comment by brett-warren-arm** [comment]
Commits mitigating the two above issues have been pushed, so this PR should now be completely ready

**Review by hanno-becker — APPROVED** [review]
Thanks @brett-warren-arm -- I confirm this works. There are probably things that can be improved here, e.g. avoiding re-calling `cmake` upon changes of `mbedtls_config.h`, or supporting plain `make`, but for our purpose this is good enough as a start.

**Pull request merged**